Fix variable leak to the enclosing scope

arnavk opened the pull request:
Passing an empty object prevents the leak of variables to the enclosing scope leaking. This leak could result in global variables being overwritten.

Note: I'm also in the process of adding a test case that demonstrates the leak and the fix. Let's not merge before that.

arnavk committed:
Fix variable leak to the enclosing scope

camertron commented:
Ahhh I've seen other libraries do this but I never knew why, thanks for fixing! Do you think we should do a minor or patch version bump for this?

arnavk committed:
Add a context argmuent to the runtime generation + add tests

arnavk commented:
Anytime! 

This is primarily a bug fix, but it's sort of significant since the leak could potentially cause problems. I suppose, a minor version number bump is fine. What do you think?

arnavk commented:
Well, I forgot to add a `context` argument in the previous commit. Added it now. From what I understand this is a more Javascript-y way of doing things.

Also, added a test. (Not sure if I did it right, though)

camertron commented:
Test looks good to me - it's testing to make sure `this.v = function() { ... }` doesn't override the previously defined value of `v`. I tried removing your fixes and running tests, and the test you wrote fails as expected :)

camertron commented:
Published v1.1.0.

arnavk commented:
@camertron the thing I was not sure about was that I declared a second `context` within a `context` block.

arnavk commented on spec/javascript_runtime_spec.rb:
```diff
@@ -204,4 +204,14 @@
   end
+
+  context 'with global variables already present' do
+    context = V8::Context.new
```
@camertron `context` within a `context`. I was wondering if there was a better way... I suppose the inner `context` occupies the block scope and things work. I suppose there is a way to use the outer `context` inside the block. No?

camertron replied:
Ah I see what you mean... Well since the plurals runtime isn't injected into the outer context until the `rt` method is called, it should be safe to use it in your test. You could even call the `rt` method when you're ready which would do the same thing as line 211.

arnavk replied:
Oh, I see!

radzinzki commented:
Oh, I see!

KL-7 commented:
Nice catch, @radzinzki! Should we update `cldr-plurals-runtime-js` version in `twitter-cldr-js`?

arnavk commented:
We should. I'll add this change in the [RBNF PR](https://github.com/twitter/twitter-cldr-js/pull/60).
